### ISSUE TITLE
After 3.243.0 Host override cannot be combined with Dualstack, FIPS, or S3 Accelerate

### ISSUE THREAD

**Issue by kerkness** [aws/aws-sdk-php#2614]
### Describe the bug

After the update to version 3.243.0 and I think the start of using Aws\EndpointV2  it would appear that the option for `use_dual_stack_endpoint` and possibly `use_fips_endpoint` are always true.   This prevents multipart uploads which use accelerator and any writes that use accelerator with possible other issues as a result.

When tracing through the updates made in 3.243.0 and subsequent updates I was unable to determine where the options are actually being set to true, or if the default behavior for these options is being assumed as true on the AWS end and explicitly setting them to false is being ignored.

I am using Laravel 9 with league/flysystem-aws-s3-v3  but the issue appears to be isolated with changes starting with aws-sdk-php v3.243.0 and later.  Explicitly setting `"aws/aws-sdk-php": "3.242.1"`  in composer resolves the issue

### Expected Behavior

Default values for  `use_dual_stack_endpoint` and `use_fips_endpoint` is false.
Setting env variables `AWS_USE_DUAL_STACK_ENDPOINT` and `AWS_USE_FIPS_ENDPOINT` to false is respected.

### Current Behavior

Default values for  `use_dual_stack_endpoint` and `use_fips_endpoint` appears to be true.
Setting env variables `AWS_USE_DUAL_STACK_ENDPOINT` and `AWS_USE_FIPS_ENDPOINT` to false is ignored.

### Reproduction Steps

Create an S3 client with default Laravel S3 config and create a multipart upload
```
 $client = Storage::disk('s3')->getClient();
 
 $result = $client->createMultipartUpload([
                'Bucket'             => $bucket,
                'Key'                => $key,
                'ContentType'        => $type,
                'ContentDisposition' => 'inline',
            ]);
```
Results in error  `Host override cannot be combined with Dualstack, FIPS, or S3 Accelerate`

Environment variables used

```
AWS_ACCESS_KEY_ID=*****
AWS_SECRET_ACCESS_KEY=*****
AWS_DEFAULT_REGION=ca-central-1
AWS_BUCKET=my-bucket-name
AWS_URL="https://s3.amazonaws.com"
AWS_POST_END_POINT="https://${AWS_BUCKET}.s3.amazonaws.com/"
AWS_USE_PATH_STYLE_ENDPOINT=false
AWS_USE_ACCELERATE_ENDPOINT=true
AWS_USE_DUAL_STACK_ENDPOINT=false
AWS_USE_FIPS_ENDPOINT=false
```

### Additional Information/Context

The specific error  `Host override cannot be combined with Dualstack, FIPS, or S3 Accelerate` does not appear to be common but  has recently been reported with aws cli on stackoverflow indicating it may actually result from a more recent change on AWS/S3 side of things.

### SDK version used

3.243.0

### Environment details (Version of PHP (`php -v`)? OS name and version, etc.)

php 8.0.12, laravel/framerwork ^9.0,  league/flysystem-aws-s3-v3 ^3.0,  aws/aws-sdk-php ^3.243.0

**Comment by stobrien89** [comment]
Hi @kerkness,

Thanks for raising this and sorry to hear about the issues.  We recently changed the way we determine service endpoints to a dynamic, per invocation resolution, which has come with a few rough edges for configurations like this.  

I think this is being caused by the `accelerate` option being set to `true`.  For Dualstack, FIPS, and S3 Accelerate, we add an identifier to the host, and because a custom host name is being provided, we're not able to ensure `accelerate` is being properly set.  I'll investigate which tweaks you'll need to make to get this working, but currently custom endpoints cannot be used with the above-mentioned options.

**Comment by stobrien89** [comment]
Hi again @kerkness,

re: your `createMultiPartUpload` example, in that particular case, would you expect the effective URI to be (with a key of `foo` for example) `https://my-bucket-name.s3-accelerate.amazonaws.com/foo?uploads`? Reason I'm asking is that seems you're using a third-party utility I'm not familiar with.

Assuming the above is true, you would not need to provide an `AWS_URL` env var to get the desired result.

**Comment by kerkness** [comment]
Thanks for diving into this @stobrien89 

The full stack I am using involves Uppy.io on the client side (in my case a react frontend with their aws-s3-multipart driver) https://uppy.io/docs/aws-s3-multipart/

That client communicates with a Laravel API which invokes the aws-sdk-php with the help of a package called `tapp/laravel-uppy-s3-multipart-upload`

The full example of process can be seen here
https://github.com/TappNetwork/laravel-uppy-s3-multipart-upload/blob/master/src/Http/Controllers/UppyS3MultipartController.php#L43-L120

To summarize, the laravel/php side of things is basically a relay for the react client which provides nice progressive details on the upload. The s3 createMultipartUpload method obtains the UploadId and hands it back to the Uppy client.

I know the .env contains both AWS_URL and AWS_POST_END_POINT  variables but off the top of my head not sure if that's just because they are part of the default Laravel .env example.   When I am back at my dev station later today I can take a deeper look at the S3 Multipart Controller and see which is actually being used.

HTH

**Comment by stobrien89** [comment]
Hi @kerkness,

Thanks for the additional info!

I don't think you need to use either `AWS_URL` or `AWS_POST_END_POINT` unless you plan to use a custom (non-AWS) endpoint.  If you're using AWS S3 and not another 3rd party object store, we'll calculate the correct endpoint for those options.

**Comment by kerkness** [comment]
HI @stobrien89 

You are correct.  We are not using a custom endpoint at all, and neither `AWS_URL` or `AWS_POST_END_POINT` is actually needed. Nor are they referenced by any of the upload controllers or framework middleware, so removing them completely from the .env did resolve the issue.   

These env variables are included in the setup instructions for the `tapp/laravel-uppy-s3-multipart-upload` package and use the default s3 endpoint as the default value.  I'll submit a pull request to that package to update the documentation and remove the default values.

I'll leave it to you to determine if the aws-sdk-php should handle this situation differently.

I appreciate you looking at this issue in detail.  Thank you for your assistance.

**Comment by stobrien89** [comment]
Hi @kerkness,

Thanks for the confirmation.  I've raised this to members of other SDK teams and we've decided that the error message provided here need to be more explicit and we should probably look at tweaking some service documentation to ensure customers don't think they should be providing custom endpoints when enabling Dualstack, FIPS, or S3 Accelerate via their client config.  In short, changes will be made upstream to account for this.

I think a PR to change their docs is a good step.  Maybe just denoting that those env vars are optional would suffice.  Closing for now, but let us know if you have any questions!
